dScripts/02_server/Map/AG/AgLaserSensorServer.cpp

```diff
@@ -2,56 +2,27 @@
 
 #include "PhantomPhysicsComponent.h"
 #include "SkillComponent.h"
-#include "EntityManager.h"
-#include "AgMonumentLaserServer.h"
-#include "EntityManager.h"
 #include "ePhysicsEffectType.h"
-#include "eReplicaComponentType.h"
 
 void AgLaserSensorServer::OnStartup(Entity* self) {
-
-	PhantomPhysicsComponent* physComp = static_cast<PhantomPhysicsComponent*>(self->GetComponent(eReplicaComponentType::PHANTOM_PHYSICS));
-	physComp->SetPhysicsEffectActive(true);
-	physComp->SetEffectType(ePhysicsEffectType::REPULSE);
-	physComp->SetDirectionalMultiplier(static_cast<float>(m_RepelForce));
-	physComp->SetDirection(NiPoint3::UNIT_Y);
-
-	m_Skill = self->GetComponent<SkillComponent>();
+	self->SetBoolean(u"active", true);
+	auto repelForce = self->GetVarAs<float>(u"repelForce");
+	if (!repelForce) repelForce = m_RepelForce;
+	auto* phantomPhysicsComponent = self->GetComponent<PhantomPhysicsComponent>();
+	if (!phantomPhysicsComponent) return;
+	phantomPhysicsComponent->SetPhysicsEffectActive(true);
+	phantomPhysicsComponent->SetEffectType(ePhysicsEffectType::REPULSE);
+	phantomPhysicsComponent->SetDirectionalMultiplier(repelForce);
+	phantomPhysicsComponent->SetDirection(NiPoint3::UNIT_Y);
 }
 
 
 void AgLaserSensorServer::OnCollisionPhantom(Entity* self, Entity* target) {
-
-	if (!m_Skill) return;
-
-
-	Entity* laser = nullptr;
-
-	for (auto script : EntityManager::Instance()->GetEntitiesByComponent(eReplicaComponentType::SCRIPT)) {
-
-		AgMonumentLaserServer* hasLaser = (AgMonumentLaserServer*)script;
-
-		if (hasLaser) {
-			const auto source = script->GetPosition();
-			const auto obj = self->GetObjectID();
-
-			if (obj == 76690936093053 && Vector3::DistanceSquared(source, NiPoint3(149.007f, 417.083f, 218.346f)) <= 1.0f) {
-				laser = script;
-				break;
-			} else if (obj == 75866302318824 && Vector3::DistanceSquared(source, NiPoint3(48.6403f, 403.803f, 196.711f)) <= 1.0f) {
-				laser = script;
-				break;
-			} else if (obj == 75866302318822 && Vector3::DistanceSquared(source, NiPoint3(19.2155f, 420.083f, 249.226f)) <= 1.0f) {
-				laser = script;
-				break;
-			} else if (obj == 75866302318823 && Vector3::DistanceSquared(source, NiPoint3(-6.61596f, 404.633f, 274.323f)) <= 1.0f) {
-				laser = script;
-				break;
-			}
-		}
-	}
-
-	if (laser != nullptr) {
-		m_Skill->CalculateBehavior(m_SkillCastID, 15714, target->GetObjectID());
-	}
+	auto active = self->GetVar<bool>(u"active");
+	if (!active) return;
+	auto skillCastID = self->GetVarAs<float>(u"skillCastID");
+	if (!skillCastID) skillCastID = m_SkillCastID;
```

Comment:
This is an integer right?  Not sure what this check is checking for unless you mean if != 0

Also why are we storing this member variable on the script?  imo the variable should either be attached only to the entity, or only to the script.  not both.  If its attached to the script it should be hard coded from the get go and if its attached to the entity, it should only be fetched through GetVar.  If going for the latter, the member variable should be removed but if going for the former, this should be done once on entity initialization.
This comment also applies for the repel force

Reply:
I read the setter backwards - but we should still compare with 0 and not check for t/f with an int.